hypothesis-python/src/hypothesis/internal/conjecture/providers.py

```diff
@@ -355,8 +367,56 @@ class HypothesisProvider(PrimitiveProvider):
 
     def __init__(self, conjecturedata: Optional["ConjectureData"], /):
         super().__init__(conjecturedata)
+        self.local_constants = _get_local_constants()
         self._random = None if self._cd is None else self._cd._random
 
+    def _maybe_draw_constant(
+        self,
+        choice_type: ChoiceTypeT,
+        constraints: ChoiceConstraintsT,
+        *,
+        p: float = 0.05,
+    ) -> Optional["ConstantT"]:
+        assert self._random is not None
+        assert choice_type != "boolean"
+
+        # check whether we even want a constant before spending time computing
+        # and caching the allowed constants.
+        if self._random.random() > p:
+            return None
+
+        key = (choice_type, choice_constraints_key(choice_type, constraints))
+        if key not in CONSTANTS_CACHE:
+            CONSTANTS_CACHE[key] = (
+                tuple(
+                    choice
+                    for choice in GLOBAL_CONSTANTS[choice_type]
```

Comment:
I think we're iterating over a set here, which means [hash randomization](https://docs.python.org/3/using/cmdline.html#envvar-PYTHONHASHSEED) will sharply diminish the usefulness of our `@seed()` / `--hypothesis-seed` reproducibility options.

Of course some diminution of determinism is unavoidable when we're pulling this trick, but I'd prefer to limit it if we can.  Perhaps `SortedSet` would be appropriate?  (with the key for floats to handle nan)